### DISCUSSION TITLE
Multithreading

### DISCUSSION THREAD

**Discussion by alphaqu** [activej/activej#194] [Ideas]
So by my testing the serializer the creation is terribly single-threaded and my application time is 50% of the creation. Please optimize/multithread the serializer creation. There is not really anything else I can put here as its pretty explanatory.

**Comment by eduard-vasinskyi** [comment]
Hello, @alphaqu
Generated serializers are fully thread-safe and can be readily used in a multithreaded environment. 
ActiveJ Serializer is highly optimized for runtime (e.g. actual data serialization/deserialization). Creating a serializer can be considered a set-up/start-up time of application and therefore does not affect runtime. 
Just make sure you  generate serializers once and then cache and reuse them rather than generating a new one each time.

**Comment by alphaqu** [comment]
I do already reuse the same serializer. and I do heavily multithread using them. The issue is that my subclasses are huge and it takes about 2 seconds to create and 2 seconds to deserialize what I need to. 50% of the time is not multithreaded which is horrible for people with cpu's like threadrippers where my program is pretty terrible as their single thread creating a serializer is really slow. My application is improving the startup time hence why I see improvements to be made.

**Comment by eduard-vasinskyi** [comment]
The main goal of [ActiveJ Serializer](https://serializer.activej.io/) is achieving maximum performance in runtime when serializing/deserializing massive data sets and unbound streams. An ideal field of application is serialization/deserialization of data sets, logs, streams, RPC requests/responses that consist of millions of typical messages. In this case, the cost of creating a serializer is [amortized to a zero](https://en.wikipedia.org/wiki/Amortized_analysis)

That is why we use advanced techniques, such as code generation and class graph introspection. In this case, the time it takes to create a serializer is not important. 

Such an approach is in the spirit of JVM - slow class and bytecode loading followed by further JIT optimizations.

Nevertheless, we'll try to see what can be done for speeding up serializer code generation.
 
_PS. If you face slow serializer creation in [DashLoader project](https://github.com/alphaqu/DashLoader) (which you mentioned in issue #72 ), I have noticed that you call `SerializerBuilder#build` from within `parallelStream()`. I believe you would get better performance if you use explicit multithreading (via `Executor`, for example) rather than counting on built-in multithreading capabilities of `parallelStream()`._
